### PR TITLE
Adds a helper for adding data points

### PR DESCRIPTION
**Pull request by ChristopherRabotin**
Implements #89 .
Subclassing this helper eases writing data points in bulk. All data points are immutable, insuring they do not get overwritten. Each subclass can write to its own database. The time series names can also be based on one or more defined fields.

Annotated example:
```python
    class MySeriesHelper(SeriesHelper):
        class Meta:
            # Meta class stores time series helper configuration.
            client = TestSeriesHelper.client
            # The client should be an instance of InfluxDBClient.
            series_name = 'events.stats.{server_name}'
            # The series name must be a string. Add dependent field names in curly brackets.
            fields = ['time', 'server_name']
            # Defines all the fields in this time series.
            bulk_size = 5
            # Defines the number of data points to store prior to writing on the wire.
    
    # The following will create *five* (immutable) data points.
    # Since bulk_size is set to 5, upon the fifth construction call, *all* data
    # points will be written on the wire via MySeriesHelper.Meta.client.
    MySeriesHelper(server_name='us.east-1', time=159)
    MySeriesHelper(server_name='us.east-1', time=158)
    MySeriesHelper(server_name='us.east-1', time=157)
    MySeriesHelper(server_name='us.east-1', time=156)
    MySeriesHelper(server_name='us.east-1', time=155)
    
    # To manually submit data points which are not yet written, call commit:
    MySeriesHelper.commit()
    
    # To inspect the JSON which will be written, call _json_body_():
    MySeriesHelper._json_body_()
```

### CODE REVIEW

File: influxdb/__init__.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 
 __all__ = [
     'InfluxDBClient',
     'DataFrameClient',
+    'SeriesHelper',
 ]
 
 
-__version__ = '0.1.13'
+__version__ = '0.1.14'
```
Comment (ChristopherRabotin) [inline influxdb/__init__.py]: Not sure if this was needed.

Comment (ReAzem) [inline influxdb/__init__.py]: No, I will bump the version myself.
I could release 0.1.14 before I merge your PR.

**Comment by ReAzem** [comment]
Hi,

I like this idea. I can see how it would make the lib very user-friendly. I have not yet looked at the code but I agree with the example code you have provided.

Thank you for your contribution :)

**Comment by ChristopherRabotin** [comment]
@ReAzem , thanks for your comments and appreciation. I'm not sure I understand what you mean by _random fixes_ though.
I'll revert the version back down to `0.1.13` and will look into why the tests failed. I only ran the test I added not the whole suite.

Let me know if there's anything I can help you with on any upcoming release. I hope to start using influxdb in prod starting the middle of next week. :D

**Comment by ChristopherRabotin** [comment]
In addition, I'm not completely familiar with the best practices for InfluxDB: if some data is stored as the series name (e.g. `server_name` in the example provided above), should it also be stored as a time series field too? It is trivial change to allow for the fields used in the series names to not be fields in the time series themselves.

**Comment by ReAzem** [comment]
> Are you referring to the requirements.txt change?[/edit]

I was refering to the .gitignore

>Let me know if there's anything I can help you with on any upcoming release. I hope to start using influxdb in prod starting the middle of next week. :D

The sooner you get the tests working, the sooner I can merge. As this seems important to you, I will do my best to get this out by the end of the week.

**Comment by ReAzem** [comment]
Can you add an ```autocommit``` parameter (bool) to the meta class? I would default it to false.

**Comment by ChristopherRabotin** [comment]
Sure. Am I right to assume that you'd like it to call `write_points` every time a single data point is created (hence similar to setting `bulk_size` to `1`)?

**Comment by ChristopherRabotin** [comment]
On another note, do you know how I can find the traceback of the failed test on Travis CI's [build](https://travis-ci.org/influxdb/influxdb-python/builds/46982815)? That would help since the test runs locally with `python -m unittest tests.influxdb.helper_test`. Thanks.

**Comment by ReAzem** [comment]
> Sure. Am I right to assume that you'd like it to call write_points every time a single data point is created (hence similar to setting bulk_size to 1)?

When autocommit is false, ```commit()``` would have to be called by the user manually. The bulk size would have no meaning.

**Comment by ReAzem** [comment]
> how I can find the traceback of the failed test on Travis CI's build?

You can find logs of your build [here](https://travis-ci.org/influxdb/influxdb-python/builds/46982815)

**Comment by ChristopherRabotin** [comment]
FYI: I pushed a new commit but it doesn't correct the bugs. I need to iron out issues with Python 3.

**Comment by ReAzem** [comment]
I think it's a little wierd to specify a class-level client attribute in the helper class. Clients are create at runtime so I think that helpers should also be created at runtime.

**Comment by ChristopherRabotin** [comment]
Fair point. Would you prefer passing the client to the `commit` function, such as the following:
```
    @classmethod
    def commit(cls, client):
        '''
        Commit everything from datapoints via the client.
        :return result of client.write_points.
        '''
        rtn = client.write_points(cls._json_body_())
        cls._reset_()
        return rtn
```

**Comment by ChristopherRabotin** [comment]
Okay, it finally builds!

**Comment by ReAzem** [comment]
@ChristopherRabotin sorry, I had missed your comment three days ago  I will take a look at the code soon.

**Comment by ReAzem** [comment]
Have you found a solution for the series name? I wouldn't include the values that make up the seires name in a time series field.

For now, I am fine with passing the client to the commit function, but how would auto commit work?

**Comment by ChristopherRabotin** [comment]
I can add the feature of not having time series values in the series name,
as soon as Monday if you like. I'll probably add a new field in the Meta
class.

With regards to auto commit, currently it will not auto commit if the
"autocommit" variable is set to false. If "bulk_size" is not set and
autocommit is, then a warning will be displayed. By default, autocommit
won't be enabled requiring users to set bulk size. Does that make sense?

**Comment by ReAzem** [comment]
> With regards to auto commit, currently it will not auto commit if the "autocommit" variable is set to false. If "bulk_size" is not set and autocommit is, then a warning will be displayed. By default, autocommit won't be enabled requiring users to set bulk size. Does that make sense?

Yep, makes sense.

I was referring to passing the client to the commit funciton. This would imply that ```client``` will no longer be a class attribute. In this case, how would auto-commit work?

File: tests/influxdb/helper_test.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                        'Invalid JSON body of time series returned from '
+                        '_json_body_ for several series names: {}.'
+                        .format(rcvd))
+        TestSeriesHelper.MySeriesHelper._reset_()
+        self.assertEqual(
+            TestSeriesHelper.MySeriesHelper._json_body_(),
+            [],
+            'Resetting helper did not empty datapoints.')
+
+    def testInvalidHelpers(self):
```
Comment (ReAzem) [inline tests/influxdb/helper_test.py]: Can this be broken down into several tests with validation of the error strings please?

**Comment by ChristopherRabotin** [comment]
If client is set as a class attribute, then autocommit works. Otherwise,
initialization will raise an exception saying that when autocommit is set
to True, then client must be set too. This allows both autocommit to work,
and passing a client to the commit function. In the latter case, you don't
need an InfluxDBClient instance when initializing the Series Helper class.

**Comment by ChristopherRabotin** [comment]
Sure, I'll break down the test. Would you like me to avoid a loop, or is a
loop with a dictionary fine?

**Comment by ReAzem** [comment]
> Sure, I'll break down the test. Would you like me to avoid a loop, or is a loop with a dictionary fine?

I think its easier to read if they are all separate tests. (one function for each test.)

**Comment by ReAzem** [comment]
> If client is set as a class attribute, then autocommit works. Otherwise, initialization will raise an exception saying that when autocommit is set to True, then client must be set too. This allows both autocommit to work, and passing a client to the commit function. In the latter case, you don't need an InfluxDBClient instance when initializing the Series Helper class.

I agree with this solution, good job :)

**Comment by ChristopherRabotin** [comment]
Cool, I'll implement that tomorrow (2149 here).

**Comment by ChristopherRabotin** [comment]
For the record, `flake8` is annoyingly rigorous. =)

**Comment by ChristopherRabotin** [comment]
Any idea how I can rebuild the docs test? It didn't build because of a connection reset by peer, which doesn't seem to come from the code itself: https://travis-ci.org/influxdb/influxdb-python/jobs/47553441 . Thanks:

**Comment by ReAzem** [comment]
I have triggered a rebuild. I don't know if you can do it, but there is a rebuild button in the travis-ci interface, just next to the ''download log'' button.

**Comment by ChristopherRabotin** [comment]
I didn't see that button. Maybe is it limited to collaborators of a given project. In any case, I'm glad to see that the build succeeded.

**Comment by ReAzem** [comment]
@ChristopherRabotin What is your todo list for this PR atm?

**Comment by ChristopherRabotin** [comment]
I think that my to do list is done. I believe you can merge this PR.

**Comment by ReAzem** [comment]
I am currently finishing up with v0.2.0, then I will proceed to your pr. (See #100)

**Comment by ReAzem** [comment]
@ChristopherRabotin I have begun merging in the series_helper branch. I have made several documentation changes so that the docstrings play nice with sphinx.

**Comment by ReAzem** [comment]
This doesn't seem to work. Any idea why?
Uncommenting the manual commit makes the test pass.
```python
    def test_auto_commit(self):
        """
        Tests that write_points is called after the right number of events
        """
        class AutoCommitTest(SeriesHelper):

            class Meta:
                series_name = 'events.stats.{server_name}'
                fields = ['time', 'server_name']
                bulk_size = 5
                client = InfluxDBClient()
                autocommit = True

        fake_write_points = mock.MagicMock()
        AutoCommitTest(server_name='us.east-1', time=159)
        AutoCommitTest._client.write_points = fake_write_points
        AutoCommitTest(server_name='us.east-1', time=158)
        AutoCommitTest(server_name='us.east-1', time=157)
        AutoCommitTest(server_name='us.east-1', time=156)
        AutoCommitTest(server_name='us.east-1', time=3443)
        AutoCommitTest(server_name='us.east-1', time=1546)
        #AutoCommitTest.commit()
        self.assertTrue(fake_write_points.called)
```

**Comment by ChristopherRabotin** [comment]
Good catch! I pinpointed the issue, and I'll get back to you in a couple of minutes with a fix. Given your changes to the branch, and how trivial the fix is, I'll just post it as a comment.

**Comment by ChristopherRabotin** [comment]
I actually pushed the change. Here's the change though.
Line 125 of `influxdb/helper.py` must be changed to the following.
```
if cls._autocommit and \
                sum(len(series) for series in cls._datapoints.values()) \
                >= cls._bulk_size:
            cls.commit()
```

**Comment by ReAzem** [comment]
Thank you =), I'll commit it on the series_helper branch.

You could have opened a PR on the branch.

**Comment by ChristopherRabotin** [comment]
Hey there, is there anything I can help you with on the series_helper branch? If so, let me know, thanks.

**Comment by ReAzem** [comment]
@ChristopherRabotin Hey! Sorry, I've been sick in the past couple days. I'll merge it  right now :)

Thank you for your contribution!

**Comment by ReAzem** [comment]
Merged in 66234914b3966dd0084ea072b2691eb8ea874f5a

**Pull request closed without merging**